### PR TITLE
support `date_trunc` with `millisecond` and `microsecond` and keep the same type as the input

### LINKED ISSUE #5750: `date_trunc` always returns `Timestamp(Nanosecond, None)`
### Describe the bug

while i try to use `date_trunc` to `Timestamp(Second, None)`, it returns `Timestamp(Nanosecond, None)`

```bash
❯ select date_trunc('minute', to_timestamp_seconds(61));
+---------------------------------------------------------+
| datetrunc(Utf8("minute"),totimestampseconds(Int64(61))) |
+---------------------------------------------------------+
| 1970-01-01T00:01:00                                     |
+---------------------------------------------------------+
1 row in set. Query took 0.003 seconds.

❯ select arrow_typeof(date_trunc('minute', to_timestamp_seconds(61)));
+----------------------------------------------------------------------+
| arrowtypeof(datetrunc(Utf8("minute"),totimestampseconds(Int64(61)))) |
+----------------------------------------------------------------------+
| Timestamp(Nanosecond, None)                                          |
+----------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

The behavior is similar while I apply `date_trunc` to Timestamp with Millis and Micros.

### To Reproduce

```bash
❯ select arrow_typeof(date_trunc('minute', to_timestamp_seconds(61)));
+----------------------------------------------------------------------+
| arrowtypeof(datetrunc(Utf8("minute"),totimestampseconds(Int64(61)))) |
+----------------------------------------------------------------------+
| Timestamp(Nanosecond, None)                                          |
+----------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

### Expected behavior

```bash
❯ select arrow_typeof(date_trunc('minute', to_timestamp_seconds(61)));
+----------------------------------------------------------------------+
| arrowtypeof(datetrunc(Utf8("minute"),totimestampseconds(Int64(61)))) |
+----------------------------------------------------------------------+
| Timestamp(Second, None)                                          |
+----------------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #5750

# Rationale for this change
`date_trunc` will return `TimestampSecond` when truncing minute.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 1]
```diff
@@ -280,10 +280,25 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
 
     Ok(match array {
         ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(v, tz_opt)) => {
-            ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(
-                (f)(*v)?,
-                tz_opt.clone(),
-            ))
+            let nano = (f)(*v)?;
+            match granularity.as_str() {
+                "minute" => {
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:285]: I wondered if we needed to apply the same treatment to `microsecond` and `millisecond`, and it appears that those granularities are not supported:

I https://github.com/apache/arrow-datafusion/blob/771c20ce2f0ade29a2d334e4e8494e9fbd7a5940/datafusion/physical-expr/src/datetime_expressions.rs#L218

(I was wondering as there are both Microsecond and Millisecond Timestamps as well)
https://github.com/apache/arrow-datafusion/blob/771c20ce2f0ade29a2d334e4e8494e9fbd7a5940/datafusion/common/src/scalar.rs#L116-L122

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H 

cc @waitingkuo

**Comment by waitingkuo** [comment]
thank you @Weijun-H 

I wonder whether we need to modify the return type as well. I'm not that familiar with this part. 
@alamb  would you mind leave some hints?

https://github.com/apache/arrow-datafusion/blob/bff92c85470e37189741e3412e35870d4a2cc15a/datafusion/expr/src/function.rs#L135-L137

**Comment by alamb** [comment]
> I wonder whether we need to modify the return type as well. I'm not that familiar with this part.


That is a good call @waitingkuo  -- I think you are correct (maybe we would have to refactor some code to handle the first argument in both places)

**Comment by Weijun-H** [comment]
> > I wonder whether we need to modify the return type as well. I'm not that familiar with this part.
> 
> That is a good call @waitingkuo -- I think you are correct (maybe we would have to refactor some code to handle the first argument in both places)

It is a good idea, but I am not sure how to achieve it.

**Comment by waitingkuo** [comment]
@Weijun-H 
you could do something like this:
```rust
    match input_expr_types[0] {
                DataType::Timestamp(TimeUnit::Nanosecond, _) => Ok(DataType::Timestamp(TimeUnit::Nanosecond, None)),
                DataType::Timestamp(TimeUnit::Microsecond, _) => Ok(DataType::Timestamp(TimeUnit::Microsecond, None)),
                DataType::Timestamp(TimeUnit::Millisecond, _) => Ok(DataType::Timestamp(TimeUnit::Millisecond, None)),
                DataType::Timestamp(TimeUnit::Second, _) => Ok(DataType::Timestamp(TimeUnit::Second, None)),
                _ => return Err(DataFusionError::Internal(
                    "The date_trunc function can only accept timestamp as the first arg.".to_string(),
                ))
            }
```

**Comment by Weijun-H** [comment]
> @Weijun-H you could do something like this:
> 
> ```rust
>     match input_expr_types[0] {
>                 DataType::Timestamp(TimeUnit::Nanosecond, _) => Ok(DataType::Timestamp(TimeUnit::Nanosecond, None)),
>                 DataType::Timestamp(TimeUnit::Microsecond, _) => Ok(DataType::Timestamp(TimeUnit::Microsecond, None)),
>                 DataType::Timestamp(TimeUnit::Millisecond, _) => Ok(DataType::Timestamp(TimeUnit::Millisecond, None)),
>                 DataType::Timestamp(TimeUnit::Second, _) => Ok(DataType::Timestamp(TimeUnit::Second, None)),
>                 _ => return Err(DataFusionError::Internal(
>                     "The date_trunc function can only accept timestamp as the first arg.".to_string(),
>                 ))
>             }
> ```

It seems that we should modify the function signature, like
```
date_trunc('Timestamp(Second, None)', to_timestamp_seconds(61))?
```

**Comment by waitingkuo** [comment]
hi @Weijun-H  sorry i think we should use `input_expr_types[1]` instead

**Comment by Weijun-H** [comment]
> hi @Weijun-H sorry i think we should use `input_expr_types[1]` instead

The return type should be based on `input_expr_types[0]`. If `input_expr_types[0]` is 'minute', the return type should be `Ok(DataType::Timestamp(TimeUnit::Second, None))`.  WDYT?

**Comment by waitingkuo** [comment]
> WDYT

i think it's better to follow the origin input timestamp type

@alamb do you think it make sense?

**Comment by alamb** [comment]
> i think it's better to follow the origin input timestamp type

Given the complexity that seems to be creeping in, I do think keeping the same type would keep things simpler

**Review by alamb — APPROVED** [review]
Looks good to me - thank you @Weijun-H

**Comment by jiacai2050** [comment]
If date_trunc keep origin datetype, doesn't this also apply to other time function? like date_bin.

**Comment by alamb** [comment]
>  If date_trunc keep origin datetype, doesn't this also apply to other time function? like date_bin.

I agree @jiacai2050

**Comment by jiacai2050** [comment]
I will open a new PR to fix other functions.

**Pull request merged**